src/fd.cpp

```diff
@@ -110,19 +159,99 @@ static int wait_thread(void *arg) {
     std::fill(args->results->begin(), args->results->end(), NA_INTEGER);
   }
 
+  return 0;
+
+}
+
+static int wait_thread_single(void *arg) {
+
+  tct_thrd_detach(tct_thrd_current());
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  if (wait_on_fds(args))
+    return 1;
+
   callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
 
   return 0;
 
 }
 
+static int wait_thread_persistent(void *arg) {
+
+  tct_thrd_detach(tct_thrd_current());
+  thread_active->store(true); // atomic allows update prior to acquiring lock
+
+  if (cv.lock()) THREAD_RETURN(1);
+  if (cv.signal()) { cv.unlock(); THREAD_RETURN(1); } // signal to sync with main thread
+  cv.busy(false);
+  while (!cv.busy()) {
+    if (cv.wait()) { cv.unlock(); THREAD_RETURN(1); }
+  }
+  if (cv.unlock()) THREAD_RETURN(1);
```

Comment:
I can't condone this code, it's way too intricate. All these `THREAD_RETURN` calls seem to me to be a good case for RAII (in which case you could use the existing ConditionVariable class).

Reply:
Given my answers below (i) that I don't care about the errors and (ii) I can standardise all the condition variable usage without the atomic, it would seem that I can indeed use the existing ConditionVariable class.